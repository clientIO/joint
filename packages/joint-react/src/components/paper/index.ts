export * from './paper';
export * from './paper.types';
export * from './render-element/paper-element-item';
export * from './render-element/paper-html-container';
