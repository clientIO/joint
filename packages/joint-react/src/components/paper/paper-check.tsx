/* eslint-disable sonarjs/cognitive-complexity */
import { useEffect } from 'react';
import type { GraphElement } from '../../types/element-types';
import type { PaperProps } from './paper';
import type { ReactPaperOptions } from '../paper-provider/paper-provider';

const PAPER_PROPS_NAMES: Array<keyof ReactPaperOptions> = [
  'afterRender',
  'allowLink',
  'anchorNamespace',
  'async',
  'autoFreeze',
  'background',
  'beforeRender',
  'cellViewNamespace',
  'clickThreshold',
  'connectionPointNamespace',
  'connectionStrategy',
  'connectorNamespace',
  'defaultAnchor',
  'defaultConnectionPoint',
  'defaultConnector',
  'defaultLink',
  'defaultLinkAnchor',
  'defaultRouter',
  'drawGrid',
  'drawGridSize',
  'elementView',
  'embeddingMode',
  'findParentBy',
  'frontParentOnly',
  'gridSize',
  'guard',
  'height',
  'highlighterNamespace',
  'highlighting',
  'interactive',
  'labelsLayer',
  'linkAnchorNamespace',
  'linkPinning',
  'linkView',
  'magnetThreshold',
  'markAvailable',
  'measureNode',
  'moveThreshold',
  'multiLinks',
  'onViewPostponed',
  'onViewUpdate',
  'overflow',
  'preventContextMenu',
  'preventDefaultBlankAction',
  'preventDefaultViewAction',
  'restrictTranslate',
  'routerNamespace',
  'snapLabels',
  'snapLinks',
  'snapLinksSelf',
  'sorting',
  'validateConnection',
  'validateEmbedding',
  'validateMagnet',
  'validateUnembedding',
  'viewport',
  'width',
];
/**
 * `VerifyProps` is a component that checks the properties of the Paper component in development mode.
 * This component is ignored in production mode.
 * @param props - The properties of the Paper component.
 * @returns - Returns null in production mode, or a verification component in development mode.
 */
export function PaperCheck<ElementItem extends GraphElement = GraphElement>(
  props: PaperProps<ElementItem>
) {
  useEffect(() => {
    if (process.env.NODE_ENV === 'production') {
      return;
    }
    const warnings = PAPER_PROPS_NAMES.filter((propertyName) => props[propertyName] !== undefined);
    if (warnings.length === 0) {
      return;
    }
    // eslint-disable-next-line no-console
    console.warn(
      `[Paper] The following props were set directly on \n"<Paper>": "${warnings.join(', ')}".
If you're using <PaperProvider>, these options should be defined there instead.
When"<PaperProvider> is present, any props set on <Paper> will be ignored.
If you're NOT using "<PaperProvider>", then it's perfectly fine to set options directly on <Paper>.`
    );
  }, [props]);
  return null;
}
