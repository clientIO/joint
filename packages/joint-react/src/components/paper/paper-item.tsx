import { type ReactNode } from 'react';
import { createPortal } from 'react-dom';
import type { CellWithId } from '../../types/cell.types';
import type { GraphElement } from '../../types/element-types';
import typedMemo from '../../utils/typed-memo';

export interface PaperPortalProps<Data extends CellWithId = GraphElement> {
  /**
   * A function that renders the element. It is called every time the element is rendered.
   */
  readonly renderElement: (element: Data) => ReactNode;
  /**
   * The cell to render.
   */
  readonly nodeSvgGElement: SVGElement;
}

/**
 * Helper paper render component wrapped in a portal.
 * This component is used to render a paper element inside a portal.
 * It takes a renderElement function, a cell, and a containerElement as props.
 * The renderElement function is called with the cell as an argument and its return value is rendered inside the containerElement.
 * @param props - The props for the component.
 * @group Components
 * @description
 * This component is used to render a paper element inside a portal.
 * @returns The rendered element inside the portal.
 * @internal
 */
function Component<Data extends CellWithId = GraphElement>(props: PaperPortalProps<Data>) {
  const { renderElement, nodeSvgGElement, ...rest } = props;
  const cell = rest as Data;
  const element = renderElement(cell);

  return createPortal(element, nodeSvgGElement);
}

export const PaperItem = typedMemo(Component);
