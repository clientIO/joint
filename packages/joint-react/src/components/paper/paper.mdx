import { Meta, Story, Canvas, Controls, Markdown } from '@storybook/blocks';
import * as Stories from './paper.stories';

<Meta of={Stories} />

# Paper Component

Paper component that renders the JointJS elements (nodes) inside HTML.
It uses `renderElement` to render the elements.

For more about native paper component, see [paper](https://docs.jointjs.com/learn/quickstart/paper).

    <Canvas of={Stories.WithRectElement} />
    <Canvas of={Stories.WithHtmlElement} />
<Controls/>
