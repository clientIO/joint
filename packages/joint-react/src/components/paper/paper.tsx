import { type dia } from '@joint/core';
import {
  useContext,
  useEffect,
  useMemo,
  useRef,
  useState,
  type CSSProperties,
  type ReactNode,
} from 'react';
import type { GraphElement } from '../../types/element-types';
import { noopSelector } from '../../utils/noop-selector';
import { useCreatePaper } from '../../hooks/use-create-paper';
import { useElements } from '../../hooks/use-elements';
import { CellIdContext } from '../../context/cell-id.context';
import { HTMLElementItem, SVGElementItem } from './paper-element-item';
import { type GraphProps } from '../graph-provider/graph-provider';
import typedMemo from '../../utils/typed-memo';
import type { PaperEvents } from '../../types/event.types';
import { REACT_TYPE } from '../../models/react-element';
import { useAreElementMeasured } from '../../hooks/use-are-elements-measured';
import { PaperHTMLContainer } from './paper-html-container';
import { useGraph } from '../../hooks';
import { PaperProvider, type ReactPaperOptions } from '../paper-provider/paper-provider';
import { PaperContext } from '../../context';
import { PaperCheck } from './paper-check';
export interface OnLoadOptions {
  readonly paper: dia.Paper;
  readonly graph: dia.Graph;
}
export type RenderElement<ElementItem extends GraphElement = GraphElement> = (
  element: ElementItem
) => ReactNode;

/**
 * The props for the Paper component. Extend the `dia.Paper.Options` interface.
 * For more information, see the JointJS documentation.
 * @see https://docs.jointjs.com/api/dia/Paper
 */
export interface PaperProps<ElementItem extends GraphElement = GraphElement>
  extends ReactPaperOptions,
    GraphProps,
    PaperEvents {
  /**
   * A function that renders the element.
   * 
   * Note: Jointjs works by default with SVG's so by default renderElement is append inside the SVGElement node.
   * To use HTML elements, you need to use the `HTMLNode` component or `foreignObject` element.
   * 
   * This is called when the data from `elementSelector` changes.
   * @example
   * Example with `global component`:
   * ```tsx
   * type BaseElementWithData = InferElement<typeof initialElements>
   * function RenderElement({ label }: BaseElementWithData) {
   *  return <HTMLElement className="node">{label}</HTMLElement>
   * }
   * ```
   * @example
   * Example with `local component`:
   * ```tsx
   * 
  type BaseElementWithData = InferElement<typeof initialElements>
  const renderElement: RenderElement<BaseElementWithData> = useCallback(
      (element) => <HTMLElement className="node">{element.label}</HTMLElement>,
      []
  )
   * ```
   */
  readonly renderElement?: RenderElement<ElementItem>;
  /**
   * Event called when all elements are properly measured (has all elements width and height greater than 1 - default).
   * In react, we cannot detect jointjs paper render:done event properly, so we use this special event to check if all elements are measured.
   * It is useful for like onLoad event to do some layout or other operations with `graph` or `paper`.
   */
  readonly onElementsSizeReady?: (options: OnLoadOptions) => void;

  /**
   * Event called when the paper is resized.
   * It is useful for like onLoad event to do some layout or other operations with `graph` or `paper`.
   */
  readonly onElementsSizeChange?: (options: OnLoadOptions) => void;

  /**
   * The style of the paper element.
   */
  readonly style?: CSSProperties;
  /**
   * Class name of the paper element.
   */
  readonly className?: string;

  /**
   * A function that selects the elements to be rendered.
   * It defaults to the `GraphElement` elements because `dia.Element` is not a valid React element (it do not change reference after update).
   * @default (item: dia.Cell) => `BaseElement`
   * @see GraphElement
   */
  readonly elementSelector?: (item: GraphElement) => ElementItem;
  /**
   * The scale of the paper. It's useful to create for example a zoom feature or minimap Paper.
   */

  readonly scale?: number;
  /**
   * Children to render. Paper automatically wrap the children with the PaperContext, if there is no PaperContext in the parent tree.
   */
  readonly children?: ReactNode;

  /**
   * On load custom element.
   * If provided, it must return valid HTML or SVG element and it will be replaced with the default paper element.
   * So it overwrite default paper rendering.
   * It is used internally for example to render `PaperScroller` from [joint plus](https://www.jointjs.com/jointjs-plus) package.
   * @param paper - The paper instance
   * @returns
   */
  readonly overwriteDefaultPaperElement?: (paper: dia.Paper) => HTMLElement | SVGElement;

  /**
   * The threshold for click events in pixels.
   * If the mouse moves more than this distance, it will be considered a drag event.
   * @default 10
   */
  readonly clickThreshold?: number;

  /**
   * Enabled if renderElements is render to pure HTML elements.
   * By default, `joint/react` renderElements to SVG elements, so for using HTML elements without this prop, you need to use `foreignObject` element.
   * @default false
   */
  readonly useHTMLOverlay?: boolean;
}

// eslint-disable-next-line jsdoc/require-jsdoc
function Component<ElementItem extends GraphElement = GraphElement>(
  props: PaperProps<ElementItem>
) {
  const {
    renderElement,
    style,
    className,
    elementSelector = noopSelector as (item: GraphElement) => ElementItem,
    scale,
    children,
    onElementsSizeReady,
    onElementsSizeChange,
    useHTMLOverlay,
    ...paperOptions
  } = props;

  const { paperContainerElement, paperCtx } = useCreatePaper({
    ...paperOptions,
    scale,
  });

  const paperContext = useContext(PaperContext);
  if (!paperContext) {
    throw new Error('Paper must be used within a `PaperProvider` or `Paper` component');
  }
  const { recordOfSVGElements } = paperContext;

  const graph = useGraph();
  const [HTMLRendererContainer, setHTMLRendererContainer] = useState<HTMLElement | null>(null);
  const elements = useElements((items) => items.map(elementSelector));
  const areElementsMeasured = useAreElementMeasured();
  // Keep previous sizes in a ref
  const previousSizesRef = useRef<number[][]>([]);

  // Whenever elements change (or we’ve just become measured) compare old ↔ new
  useEffect(() => {
    if (!paperCtx) return;
    if (!onElementsSizeChange) return;
    if (!areElementsMeasured) return;
    const { paper } = paperCtx;
    if (!paper) return;

    // Build current list of [width, height]
    const currentSizes = elements.map(({ width = 0, height = 0 }) => [width, height]);
    const previousSizes = previousSizesRef.current;
    let changed = false;

    // Quick bail-out on length mismatch
    if (previousSizes.length === currentSizes.length) {
      // Otherwise scan for any width/height diff
      for (const [index, currentSize] of currentSizes.entries()) {
        if (
          previousSizes[index][0] !== currentSize[0] ||
          previousSizes[index][1] !== currentSize[1]
        ) {
          changed = true;
          break;
        }
      }
    } else {
      changed = true;
    }

    if (!changed) {
      return;
    }
    // store for next time
    previousSizesRef.current = currentSizes;
    onElementsSizeChange({ paper, graph: paper.model });
  }, [elements, areElementsMeasured, onElementsSizeChange, paperCtx]);

  const hasRenderElement = !!renderElement;

  const paperContainerStyle = useMemo(
    (): CSSProperties => ({
      opacity: areElementsMeasured ? 1 : 0,
      pointerEvents: areElementsMeasured ? 'all' : 'none',
      position: 'relative',
      overflow: 'hidden',
      width: '100%',
      height: '100%',
      ...style,
    }),
    [areElementsMeasured, style]
  );

  const measured = useRef(false);

  useEffect(() => {
    if (!paperCtx) {
      return;
    }
    if (measured.current) {
      // If we already measured, we can skip this effect
      return;
    }
    const { paper } = paperCtx;
    if (!paper) {
      return;
    }
    if (areElementsMeasured) {
      measured.current = true;
      return onElementsSizeReady?.({ paper, graph: paper.model });
    }

    // Handling dev warning check
    if (process.env.NODE_ENV !== 'production') {
      const timeout = setTimeout(() => {
        if (!areElementsMeasured) {
          // eslint-disable-next-line no-console
          console.error(
            'The elements are not measured yet, please check if elements has defined width and height inside the nodes or using `MeasuredNode` component.'
          );
        }
      }, 1000);
      return () => {
        clearTimeout(timeout);
      };
    }
  }, [areElementsMeasured, graph, onElementsSizeReady, paperCtx]);

  const content = (
    <>
      {hasRenderElement && useHTMLOverlay && (
        <PaperHTMLContainer onSetElement={setHTMLRendererContainer} />
      )}
      {hasRenderElement &&
        elements.map((cell) => {
          if (!cell.id) {
            return null;
          }
          const portalHTMLElement = recordOfSVGElements[cell.id];
          if (!portalHTMLElement) {
            return null;
          }
          if (cell.type !== REACT_TYPE) {
            return null;
          }

          return (
            <CellIdContext.Provider key={cell.id} value={cell.id}>
              {useHTMLOverlay && HTMLRendererContainer ? (
                <HTMLElementItem
                  {...cell}
                  portalElement={HTMLRendererContainer}
                  renderElement={renderElement}
                />
              ) : (
                <SVGElementItem
                  {...cell}
                  portalElement={portalHTMLElement}
                  renderElement={renderElement}
                />
              )}
            </CellIdContext.Provider>
          );
        })}
    </>
  );

  if (paperCtx) {
    // we need this for shared paper context - joint plus
    paperCtx.renderElement = renderElement as RenderElement<GraphElement>;
  }
  const hasPaper = !!paperCtx?.paper;

  return (
    <>
      <div className={className} ref={paperContainerElement} style={paperContainerStyle}>
        {hasPaper && content}
      </div>
      {hasPaper && children}
    </>
  );
}
// eslint-disable-next-line jsdoc/require-jsdoc
function PaperWithProviders<ElementItem extends GraphElement = GraphElement>(
  props: PaperProps<ElementItem>
) {
  const hasPaperCtx = !!useContext(PaperContext);
  const { children, ...rest } = props;
  const content = <Component {...rest}>{children}</Component>;
  if (hasPaperCtx) {
    const verifyProps = process.env.NODE_ENV !== 'production' && <PaperCheck {...rest} />;
    // If PaperContext is already provided, we don't need to wrap it again
    return (
      <>
        {verifyProps}
        {content}
      </>
    );
  }
  return <PaperProvider {...rest}>{content}</PaperProvider>;
}

/**
 * Paper component that renders the JointJS paper elements inside HTML.
 * It uses `renderElement` to render the elements.
 * It must be used within a `GraphProvider` context.
 * @see GraphProvider
 * @see PaperProps
 * 
 * Props also extends `dia.Paper.Options` interface.
 * @see dia.Paper.Options 
 * @group Components
 * @example
 * Example with `global renderElement component`:
 * ```tsx
 * import { createElements, InferElement, GraphProvider, Paper } from '@joint/react'
 *
 * const initialElements = createElements([ { id: '1', label: 'Node 1' , x: 100, y: 0, width: 100, height: 50 } ])
 * type BaseElementWithData = InferElement<typeof initialElements>
 *
 * function RenderElement({ label }: BaseElementWithData) {
 *  return <HTMLElement className="node">{label}</HTMLElement>
 * }
 * function MyApp() {
 *  return <GraphProvider initialElements={initialElements}>
 *    <Paper renderElement={RenderElement} />
 *  </GraphProvider>
 * }
 * ```
 * @example
 * Example with `local renderElement component`:
 * ```tsx
  const initialElements = createElements([
    { id: '1', label: 'Node 1', x: 100, y: 0, width: 100, height: 50 },
  ])
  type BaseElementWithData = InferElement<typeof initialElements>
 
  function MyApp() {
    const renderElement: RenderElement<BaseElementWithData> = useCallback(
      (element) => <HTMLElement className="node">{element.label}</HTMLElement>,
      []
    )
 
    return (
      <GraphProvider initialElements={initialElements}>
        <Paper renderElement={renderElement} />
      </GraphProvider>
    )
  }
 * ```
 */
export const Paper = typedMemo(PaperWithProviders);
