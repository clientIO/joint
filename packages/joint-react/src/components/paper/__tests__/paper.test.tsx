/* eslint-disable react-perf/jsx-no-new-function-as-prop */
import { render, screen, waitFor } from '@testing-library/react';
import { GraphProvider } from '../../graph-provider/graph-provider';
import { createElements, type InferElement } from '../../../utils/create';
import { Paper } from '../paper';
import { MeasuredNode } from '../../measured-node/measured-node';
import { runStorybookSnapshot } from '../../../utils/run-storybook-snapshot';
import * as stories from '../paper.stories';
import { usePaper } from '../../../hooks';
import { useEffect } from 'react';

const initialElements = createElements([
  { id: '1', label: 'Node 1' },
  { id: '2', label: 'Node 2' },
]);

type Element = InferElement<typeof initialElements>;
const PAPER_WIDTH = 200;

// we need to mock `new ResizeObserver`, to return the size width 50 and height 50 for test purposes
// Mock ResizeObserver to return a size with width 50 and height 50
jest.mock('../../../utils/create-element-size-observer', () => ({
  createElementSizeObserver: jest.fn((element, onResize) => {
    // Simulate a resize event with specific width and height
    onResize({ width: 50, height: 50 });
    // Return a cleanup function that just calls `disconnect` (this is just a placeholder)
    return () => {};
  }),
}));

// Mock `useAreElementMeasured` to simulate elements being measured
jest.mock('../../../hooks/use-are-elements-measured', () => ({
  useAreElementMeasured: jest.fn(() => true),
}));

runStorybookSnapshot({
  Component: Paper,
  name: 'Paper',
  stories,
});

describe('Paper Component', () => {
  it('renders elements correctly with correct measured node and onMeasured event', async () => {
    const onMeasuredMock = jest.fn();
    let size = { width: 0, height: 0 };
    render(
      <GraphProvider initialElements={initialElements}>
        <Paper<Element>
          width={PAPER_WIDTH}
          height={150}
          onElementsSizeReady={onMeasuredMock}
          renderElement={({ label, width, height }) => {
            size = { width, height };
            return (
              <foreignObject width={width} height={height}>
                <MeasuredNode>
                  <div className="node">{label}</div>
                </MeasuredNode>
              </foreignObject>
            );
          }}
        />
      </GraphProvider>
    );
    await waitFor(() => {
      expect(screen.getByText('Node 1')).toBeInTheDocument();
      expect(screen.getByText('Node 2')).toBeInTheDocument();
      expect(onMeasuredMock).toHaveBeenCalledTimes(1);
      expect(size).toEqual({ width: 50, height: 50 });
    });
  });

  it('renders elements correctly with useHTMLOverlay enabled', async () => {
    render(
      <GraphProvider initialElements={initialElements}>
        <Paper<Element>
          useHTMLOverlay
          renderElement={({ label }) => <div className="html-node">{label}</div>}
        />
      </GraphProvider>
    );
    await waitFor(() => {
      expect(screen.getByText('Node 1')).toBeInTheDocument();
      expect(screen.getByText('Node 2')).toBeInTheDocument();
      expect(screen.getByText('Node 1').closest('.html-node')).toBeTruthy();
    });
  });

  it('calls onElementsSizeChange when element sizes change', async () => {
    const onElementsSizeChangeMock = jest.fn();
    const updatedElements = createElements([
      { id: '1', label: 'Node 1', width: 100, height: 50 },
      { id: '2', label: 'Node 2', width: 150, height: 75 },
    ]);

    const { rerender } = render(
      <GraphProvider initialElements={initialElements}>
        <Paper<Element>
          onElementsSizeChange={onElementsSizeChangeMock}
          renderElement={({ label }) => <div className="node">{label}</div>}
        />
      </GraphProvider>
    );

    // Simulate element size change by rerendering with updated elements
    rerender(
      <GraphProvider initialElements={updatedElements}>
        <Paper<Element>
          onElementsSizeChange={onElementsSizeChangeMock}
          renderElement={({ label }) => <div className="node">{label}</div>}
        />
      </GraphProvider>
    );

    await waitFor(() => {
      expect(onElementsSizeChangeMock).toHaveBeenCalledTimes(1);
    });
  });

  it('overwrites default paper element with overwriteDefaultPaperElement', () => {
    const customElement = document.createElement('div');
    customElement.className = 'custom-paper-element';
    render(
      <GraphProvider initialElements={initialElements}>
        <Paper<Element> overwriteDefaultPaperElement={() => customElement} />
      </GraphProvider>
    );
    expect(document.querySelector('.custom-paper-element')).toBeInTheDocument();
  });
  it('should fire custom event on the paper', async () => {
    const handleCustomEvent = jest.fn();

    // eslint-disable-next-line unicorn/consistent-function-scoping
    function FireEvent() {
      const paper = usePaper();
      useEffect(() => {
        paper.trigger('MyCustomEventOnClick', { message: 'Hello from custom event!' });
      }, [paper]);
      return null;
    }
    render(
      <GraphProvider initialElements={initialElements}>
        <Paper<Element> onCustomEvent={handleCustomEvent}>
          <FireEvent />
        </Paper>
      </GraphProvider>
    );

    await waitFor(() => {
      expect(handleCustomEvent).toHaveBeenCalledTimes(1);
    });
  });

  it('applies default clickThreshold and custom clickThreshold', () => {
    render(
      <GraphProvider initialElements={initialElements}>
        <Paper<Element> />
      </GraphProvider>
    );
    const paperElement = document.querySelector('.joint-paper');
    expect(paperElement).toBeInTheDocument();

    render(
      <GraphProvider initialElements={initialElements}>
        <Paper<Element> clickThreshold={20} />
      </GraphProvider>
    );
    // Ensure no errors occur when custom clickThreshold is applied
    expect(paperElement).toBeInTheDocument();
  });

  it('applies scale to the paper', () => {
    render(
      <GraphProvider initialElements={initialElements}>
        <Paper<Element> scale={2} />
      </GraphProvider>
    );
    const layersGroup = document.querySelector('.joint-layers');
    expect(layersGroup).toHaveAttribute('transform', 'matrix(2,0,0,2,0,0)');
  });

  it('uses default elementSelector and custom elementSelector', async () => {
    const customSelector = jest.fn((item) => ({ ...item, custom: true }));
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    function RenderElement({ custom }: any) {
      return <rect id={custom ? 'isCustom' : 'nope'} width={50} height={50} fill="blue" />;
    }
    render(
      <GraphProvider initialElements={initialElements}>
        <Paper<Element> elementSelector={customSelector} renderElement={RenderElement} />
      </GraphProvider>
    );

    // Ensure the customSelector is called for each element
    expect(customSelector).toHaveBeenCalledTimes(initialElements.length);

    await waitFor(() => {
      // Validate that the elements are rendered correctly
      const element = document.querySelector('#isCustom');
      expect(element).toBeInTheDocument();
      expect(element).toHaveAttribute('width', '50');
    });
  });

  it('calls onElementsSizeReady when elements are measured', async () => {
    const onElementsSizeReadyMock = jest.fn();
    render(
      <GraphProvider initialElements={initialElements}>
        <Paper<Element> onElementsSizeReady={onElementsSizeReadyMock} />
      </GraphProvider>
    );
    await waitFor(() => {
      expect(onElementsSizeReadyMock).toHaveBeenCalledTimes(1);
    });
  });
});
