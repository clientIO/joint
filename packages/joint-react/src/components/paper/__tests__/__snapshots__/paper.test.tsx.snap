// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paper "WithAutoFitContent": Paper-WithAutoFitContent 1`] = `"<div class="border-1 border-gray-300 rounded-lg shadow-md overflow-hidden p-2 mr-2" style="opacity: 1; pointer-events: all; position: relative; overflow: hidden; width: 100%; height: 100%;"><div class="joint-border-1 joint-border-gray-300 joint-rounded-lg joint-shadow-md joint-overflow-hidden joint-p-2 joint-mr-2 joint-theme-default" style="position: relative; width: 100%; height: 600px;"><div style="position: absolute; inset: 0;" class="joint-paper-background" joint-selector="background"></div><svg width="100%" height="100%" xmlns:xlink="http://www.w3.org/1999/xlink" style="position: absolute; inset: 0; overflow: hidden;" joint-selector="svg" id="v-*"><style id="v-*" type="text/css"></style><defs joint-selector="defs"></defs><g class="joint-layers" joint-selector="layers"><g style="pointer-events: none;" class="joint-grid-layer"></g><g class="joint-back-layer"></g><g class="joint-cells-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-labels-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-front-layer"></g><g class="joint-tools-layer"></g></g></svg></div></div>"`;

exports[`Paper "WithCustomEvent": Paper-WithCustomEvent 1`] = `"<div class="border-1 border-gray-300 rounded-lg shadow-md overflow-hidden p-2 mr-2" style="opacity: 1; pointer-events: all; position: relative; overflow: hidden; width: 100%; height: 100%;"><div class="joint-border-1 joint-border-gray-300 joint-rounded-lg joint-shadow-md joint-overflow-hidden joint-p-2 joint-mr-2 joint-theme-default" style="position: relative; width: 100%; height: 600px;"><div style="position: absolute; inset: 0;" class="joint-paper-background" joint-selector="background"></div><svg width="100%" height="100%" xmlns:xlink="http://www.w3.org/1999/xlink" style="position: absolute; inset: 0; overflow: hidden;" joint-selector="svg" id="v-*"><style id="v-*" type="text/css"></style><defs joint-selector="defs"></defs><g class="joint-layers" joint-selector="layers"><g style="pointer-events: none;" class="joint-grid-layer"></g><g class="joint-back-layer"></g><g class="joint-cells-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-labels-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-front-layer"></g><g class="joint-tools-layer"></g></g></svg></div></div>"`;

exports[`Paper "WithEvent": Paper-WithEvent 1`] = `"<div class="border-1 border-gray-300 rounded-lg shadow-md overflow-hidden p-2 mr-2" style="opacity: 1; pointer-events: all; position: relative; overflow: hidden; width: 100%; height: 100%;"><div class="joint-border-1 joint-border-gray-300 joint-rounded-lg joint-shadow-md joint-overflow-hidden joint-p-2 joint-mr-2 joint-theme-default" style="position: relative; width: 100%; height: 600px;"><div style="position: absolute; inset: 0;" class="joint-paper-background" joint-selector="background"></div><svg width="100%" height="100%" xmlns:xlink="http://www.w3.org/1999/xlink" style="position: absolute; inset: 0; overflow: hidden;" joint-selector="svg" id="v-*"><style id="v-*" type="text/css"></style><defs joint-selector="defs"></defs><g class="joint-layers" joint-selector="layers"><g style="pointer-events: none;" class="joint-grid-layer"></g><g class="joint-back-layer"></g><g class="joint-cells-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-labels-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-front-layer"></g><g class="joint-tools-layer"></g></g></svg></div></div>"`;

exports[`Paper "WithGrid": Paper-WithGrid 1`] = `"<div class="border-1 border-gray-300 rounded-lg shadow-md overflow-hidden p-2 mr-2" style="opacity: 1; pointer-events: all; position: relative; overflow: hidden; width: 100%; height: 100%;"><div class="joint-border-1 joint-border-gray-300 joint-rounded-lg joint-shadow-md joint-overflow-hidden joint-p-2 joint-mr-2 joint-theme-default" style="position: relative; width: 100%; height: 600px;"><div style="position: absolute; inset: 0;" class="joint-paper-background" joint-selector="background"></div><svg width="100%" height="100%" xmlns:xlink="http://www.w3.org/1999/xlink" style="position: absolute; inset: 0; overflow: hidden;" joint-selector="svg" id="v-*"><style id="v-*" type="text/css"></style><defs joint-selector="defs"></defs><g class="joint-layers" joint-selector="layers"><g style="pointer-events: none;" class="joint-grid-layer"><svg xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" id="v-*" width="0" height="0" x="0" y="600"><defs id="v-*"><pattern id="pattern_view32_0" patternUnits="userSpaceOnUse" width="10" height="10" x="0" y="-600"><rect id="v-*" width="1" height="1" fill="#AAAAAA"></rect></pattern></defs><rect id="v-*" width="100%" height="100%" fill="url(#pattern_view32_0)"></rect></svg></g><g class="joint-back-layer"></g><g class="joint-cells-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-labels-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-front-layer"></g><g class="joint-tools-layer"></g></g></svg></div></div>"`;

exports[`Paper "WithHTMLElement": Paper-WithHTMLElement 1`] = `"<div class="border-1 border-gray-300 rounded-lg shadow-md overflow-hidden p-2 mr-2" style="opacity: 1; pointer-events: all; position: relative; overflow: hidden; width: 100%; height: 100%;"><div class="joint-border-1 joint-border-gray-300 joint-rounded-lg joint-shadow-md joint-overflow-hidden joint-p-2 joint-mr-2 joint-theme-default" style="position: relative; width: 100%; height: 600px;"><div style="position: absolute; inset: 0;" class="joint-paper-background" joint-selector="background"></div><svg width="100%" height="100%" xmlns:xlink="http://www.w3.org/1999/xlink" style="position: absolute; inset: 0; overflow: hidden;" joint-selector="svg" id="v-*"><style id="v-*" type="text/css"></style><defs joint-selector="defs"></defs><g class="joint-layers" joint-selector="layers"><g style="pointer-events: none;" class="joint-grid-layer"></g><g class="joint-back-layer"></g><g class="joint-cells-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-labels-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-front-layer"></g><g class="joint-tools-layer"></g></g></svg></div></div>"`;

exports[`Paper "WithLinkTools": Paper-WithLinkTools 1`] = `"<div class="border-1 border-gray-300 rounded-lg shadow-md overflow-hidden p-2 mr-2" style="opacity: 1; pointer-events: all; position: relative; overflow: hidden; width: 100%; height: 100%;"><div class="joint-border-1 joint-border-gray-300 joint-rounded-lg joint-shadow-md joint-overflow-hidden joint-p-2 joint-mr-2 joint-theme-default" style="position: relative; width: 100%; height: 600px;"><div style="position: absolute; inset: 0;" class="joint-paper-background" joint-selector="background"></div><svg width="100%" height="100%" xmlns:xlink="http://www.w3.org/1999/xlink" style="position: absolute; inset: 0; overflow: hidden;" joint-selector="svg" id="v-*"><style id="v-*" type="text/css"></style><defs joint-selector="defs"></defs><g class="joint-layers" joint-selector="layers"><g style="pointer-events: none;" class="joint-grid-layer"></g><g class="joint-back-layer"></g><g class="joint-cells-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-labels-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-front-layer"></g><g class="joint-tools-layer"></g></g></svg></div></div>"`;

exports[`Paper "WithRectElement": Paper-WithRectElement 1`] = `"<div class="border-1 border-gray-300 rounded-lg shadow-md overflow-hidden p-2 mr-2" style="opacity: 1; pointer-events: all; position: relative; overflow: hidden; width: 100%; height: 100%;"><div class="joint-border-1 joint-border-gray-300 joint-rounded-lg joint-shadow-md joint-overflow-hidden joint-p-2 joint-mr-2 joint-theme-default" style="position: relative; width: 100%; height: 600px;"><div style="position: absolute; inset: 0;" class="joint-paper-background" joint-selector="background"></div><svg width="100%" height="100%" xmlns:xlink="http://www.w3.org/1999/xlink" style="position: absolute; inset: 0; overflow: hidden;" joint-selector="svg" id="v-*"><style id="v-*" type="text/css"></style><defs joint-selector="defs"></defs><g class="joint-layers" joint-selector="layers"><g style="pointer-events: none;" class="joint-grid-layer"></g><g class="joint-back-layer"></g><g class="joint-cells-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-labels-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-front-layer"></g><g class="joint-tools-layer"></g></g></svg></div></div>"`;

exports[`Paper "WithScaleDown": Paper-WithScaleDown 1`] = `"<div class="border-1 border-gray-300 rounded-lg shadow-md overflow-hidden p-2 mr-2" style="opacity: 1; pointer-events: all; position: relative; overflow: hidden; width: 100%; height: 100%;"><div class="joint-border-1 joint-border-gray-300 joint-rounded-lg joint-shadow-md joint-overflow-hidden joint-p-2 joint-mr-2 joint-theme-default" style="position: relative; width: 100%; height: 600px;"><div style="position: absolute; inset: 0;" class="joint-paper-background" joint-selector="background"></div><svg width="100%" height="100%" xmlns:xlink="http://www.w3.org/1999/xlink" style="position: absolute; inset: 0; overflow: hidden;" joint-selector="svg" id="v-*"><style id="v-*" type="text/css"></style><defs joint-selector="defs"></defs><g class="joint-layers" joint-selector="layers" transform="matrix(0.7,0,0,0.7,0,0)"><g style="pointer-events: none;" class="joint-grid-layer"></g><g class="joint-back-layer"></g><g class="joint-cells-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-labels-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-front-layer"></g><g class="joint-tools-layer"></g></g></svg></div></div>"`;
